### PR TITLE
Bump `elliptic-curve` to v0.13.0-pre.1

### PR DESCRIPTION
**Pull request by tarcieri**
Bumps `elliptic-curve` to the latest prerelease, with `ecdsa` v0.16.0-pre sourced from git.

### CODE REVIEW

**Comment by tarcieri** [comment]
@fjarri @str4d could definitely do with a cross-check on this as it contains the `ff`/`group` v0.13 upgrade as well (which I need to get into the commit message).

Some of the constants are still TODOs which I can take a look at soon.

File: k256/src/arithmetic/field.rs [hunk 1]
```diff
@@ -272,7 +236,72 @@ impl FieldElement {
...
+}
+
+impl PrimeField for FieldElement {
+    type Repr = FieldBytes;
+
+    const MODULUS: &'static str =
+        "fffffffffffffffffffffffffffffffffffffffffffffffffffffffefffffc2f";
+    const NUM_BITS: u32 = 256;
+    const CAPACITY: u32 = 255; // TODO(tarcieri): double check this
+    const TWO_INV: Self = Self::ZERO; // TODO
```
Comment (fjarri) [inline k256/src/arithmetic/field.rs:288]: I calculated these constants a while ago, you can use https://github.com/fjarri/elliptic-curves/commit/d7de9bfb7dbe0e3ddb9fad45506564fb1c8f511f as a reference (or as a test)

File: k256/src/arithmetic/field.rs [hunk 2]
```diff
@@ -290,14 +319,23 @@ impl ConstantTimeEq for FieldElement {
...
-        Self::zero()
+        Self::ZERO
     }
 }
 
 impl DefaultIsZeroes for FieldElement {}
 
 impl Eq for FieldElement {}
 
+impl From<u64> for FieldElement {
```
Comment (fjarri) [inline k256/src/arithmetic/field.rs:330]: This part I implemented a little differently in the commit above, by adding methods to specific FieldElement implementations.

File: k256/src/arithmetic/field.rs [hunk 3]
```diff
@@ -420,9 +458,32 @@ impl Neg for &FieldElement {
     }
 }
 
+impl Sum for FieldElement {
+    fn sum<I: Iterator<Item = Self>>(iter: I) -> Self {
+        iter.fold(Self::ZERO, |acc, w| acc + w)
```
Comment (fjarri) [inline k256/src/arithmetic/field.rs:463]: `iter.reduce(core::ops::Add::add).unwrap_or(Self::ZERO)` makes it use one less addition. Same for the others.

File: k256/src/arithmetic/scalar.rs [hunk 4]
```diff
@@ -290,22 +284,40 @@ impl Field for Scalar {
...
+    }
 }
 
 impl PrimeField for Scalar {
     type Repr = FieldBytes;
 
+    const MODULUS: &'static str = ORDER_HEX;
     const NUM_BITS: u32 = 256;
     const CAPACITY: u32 = 255;
+    const TWO_INV: Self = Self::ZERO; // TODO
```
Comment (fjarri) [inline k256/src/arithmetic/scalar.rs:313]: See the commit linked above for the values of the constants.

File: k256/src/arithmetic/scalar.rs [hunk 5]
```diff
@@ -574,34 +588,58 @@ impl MulAssign<&Scalar> for Scalar {
...
 impl ReduceNonZero<U512> for Scalar {
-    fn from_uint_reduced_nonzero(w: U512) -> Self {
+    fn reduce_nonzero(w: U512) -> Self {
         WideScalar(w).reduce_nonzero()
     }
 }
 
+impl Sum for Scalar {
+    fn sum<I: Iterator<Item = Self>>(iter: I) -> Self {
+        iter.fold(Self::ZERO, |acc, w| acc + w)
```
Comment (fjarri) [inline k256/src/arithmetic/scalar.rs:621]: See the comment above about `fold` vs `reduce`

File: k256/src/arithmetic/field.rs (hunk 1 above)
Comment (tarcieri) [reply k256/src/arithmetic/field.rs:288]: Aah right, thanks!

Sidebar: in `p256` I was able to constify `FieldElement` enough to calculate them at compile time.

File: k256/src/arithmetic/field.rs (hunk 3 above)
Comment (tarcieri) [reply k256/src/arithmetic/field.rs]: Pushed improved `Sum`/`Product` impls in https://github.com/RustCrypto/elliptic-curves/pull/730/commits/d26419152b348fb13aa2338bb0b1107165218aff.

I used `.copied()` for when `Item` is borrowed... not sure if you had a better solution than that.

Comment (fjarri) [reply k256/src/arithmetic/field.rs]: I used `cloned()` which I guess amounts to the same thing.

Comment (tarcieri) [reply k256/src/arithmetic/field.rs]: Yeah saw your version now... nice calling the other impl. Did that in [a10b1b8](https://github.com/RustCrypto/elliptic-curves/pull/730/commits/a10b1b81b3cacb179cbdc1d8b39bf0f8b887e61d)

File: k256/src/arithmetic/field.rs (hunk 1 above)
Comment (tarcieri) [reply k256/src/arithmetic/field.rs]: Added in [1040324](https://github.com/RustCrypto/elliptic-curves/pull/730/commits/104032401b9411086467e99e4f29fd32588fc83e)

File: k256/src/arithmetic/field.rs (hunk 2 above)
Comment (tarcieri) [reply k256/src/arithmetic/field.rs:342]: Used your implementation in [1040324](https://github.com/RustCrypto/elliptic-curves/pull/730/commits/104032401b9411086467e99e4f29fd32588fc83e)

File: k256/src/arithmetic/scalar.rs (hunk 5 above)
Comment (tarcieri) [reply k256/src/arithmetic/scalar.rs]: Done

**Comment by tarcieri** [comment]
In the interest of unblocking other work, I'd like to get this merged ASAP.

I made a tracking ticket for the constants which are presently TODOs: #731

They should be pretty easy to calculate once I get sage reinstalled.

**Comment by fjarri** [comment]
> They should be pretty easy to calculate once I get sage reinstalled.

I used the online version to calculate the constants for `k256` because it's such a pain to install

**Comment by fjarri** [comment]
Speaking of `k256` constants, ideally someone should calculate them independently and cross-check.

**Comment by tarcieri** [comment]
I can also cross-check when I figure out a sage solution. Will try the online version if I can't get it working locally.

**Pull request merged**